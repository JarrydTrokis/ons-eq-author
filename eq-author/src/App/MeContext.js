import React, { createContext, useState } from "react";
import PropType from "prop-types";
import { Query, withApollo } from "react-apollo";
import { withRouter } from "react-router";
import gql from "graphql-tag";
import { get, flowRight } from "lodash";
import auth from "components/Auth";
import config from "config";
import CustomPropTypes from "custom-prop-types";
import {
  setSentryUser,
  setSentryTag,
  sendSentryError,
} from "../apollo/sentryUtils";

const createSignIn = (setHasAccessToken, history) => user => {
  localStorage.setItem("accessToken", user.ra);
  localStorage.setItem("refreshToken", user.refreshToken);

  return window
    .fetch("/signIn", {
      method: "POST",
      headers: { authorization: `Bearer ${user.ra}` },
    })
    .then(res => {
      if (!res.ok) {
        throw Error(`Server responded with a ${res.status} code.`);
      }
      setHasAccessToken(true);
      history.push(get(history, "location.state.returnURL") || "/");
    })
    .catch(e => {
      localStorage.removeItem("accessToken");
      localStorage.removeItem("refreshToken");
      auth.signOut();
      setSentryUser(user.ra);
      setSentryTag("Signing in error");
      sendSentryError(e);
      throw e;
    });
};

const createSignOut = (setHasAccessToken, { history, client }) => () => {
  if (config.REACT_APP_FULLSTORY_ORG) {
    window.FS.identify(false);
  }
  setHasAccessToken(false);
  client.clearStore();
  auth.signOut();
  localStorage.removeItem("accessToken");
  localStorage.removeItem("refreshToken");
  history.push("/sign-in");
};

export const CURRENT_USER_QUERY = gql`
  query GetMe {
    me {
      id
      displayName
      picture
      email
    }
  }
`;

export const MeContext = createContext();

const FragmentWithChildren = ({ children }) => children({});

const ContextProvider = props => {
  const accessToken = localStorage.getItem("accessToken");
  const [hasAccessToken, setHasAccessToken] = useState(Boolean(accessToken));
  const signIn = createSignIn(setHasAccessToken, props.history);
  const signOut = createSignOut(setHasAccessToken, props);
  const QueryOrFragment = hasAccessToken ? Query : FragmentWithChildren;
  return (
    <QueryOrFragment query={CURRENT_USER_QUERY}>
      {innerProps => {
        const me = get(innerProps, "data.me");
        return (
          <MeContext.Provider
            value={{
              me,
              signIn,
              hasAccessToken,
              signOut,
              awaitingUserQuery: innerProps.loading,
            }}
          >
            {props.children}
          </MeContext.Provider>
        );
      }}
    </QueryOrFragment>
  );
};

ContextProvider.propTypes = {
  history: CustomPropTypes.history.isRequired,
  children: PropType.node,
};

export const MeProvider = flowRight(
  withApollo,
  withRouter
)(ContextProvider);

export const withMe = Component => props => (
  <MeContext.Consumer>
    {({ me, signIn, hasAccessToken, awaitingUserQuery, signOut }) => (
      <Component
        {...props}
        me={me}
        signIn={signIn}
        signOut={signOut}
        hasAccessToken={hasAccessToken}
        awaitingUserQuery={awaitingUserQuery}
      />
    )}
  </MeContext.Consumer>
);
