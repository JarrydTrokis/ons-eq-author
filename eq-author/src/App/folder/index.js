import React from "react";

import { Route, Redirect } from "react-router-dom";
import { Query } from "react-apollo";
import SkipLogic from "App/shared/Logic/SkipLogic";

import Design from "./Design";
import RoutingPage from "./Routing";

import Error from "components/Error";

import GET_FOLDER_QUERY from "./graphql/getFolderQuery.graphql";

export default [
  <Route
    key="folder-design"
    path="/q/:questionnaireId/folder/:folderId/design"
    component={Design}
  />,
  <Route key="folder-logic" path="/q/:questionnaireId/folder/:folderId/logic">
    <Redirect to="routing" />
  </Route>,
  <Route
    key="folder-logic-routing"
    path="/q/:questionnaireId/folder/:folderId/routing"
    render={(props) => (
      <Query
        query={GET_FOLDER_QUERY}
        variables={{
          input: { folderId: props.match.params.folderId },
        }}
      >
        {({ loading, error, data }) => {
          if (loading) {
            return <React.Fragment />;
          }

          if (error) {
            return <Error>Error fetching folder</Error>;
          }

          if (data) {
            const { folder } = data;
            return folder.listId ? (
              <Redirect to="design" />
            ) : (
              <RoutingPage {...props} />
            );
          }
        }}
      </Query>
    )}
  />,
  <Route
    key="folder-logic-skip"
    path="/q/:questionnaireId/folder/:folderId/skip"
    render={(props) => (
      <Query
        query={GET_FOLDER_QUERY}
        variables={{
          input: { folderId: props.match.params.folderId },
        }}
      >
        {({ loading, error, data }) => {
          if (loading) {
            return <React.Fragment />;
          }

          if (error) {
            return <Error>Error fetching folder</Error>;
          }

          if (data) {
            const { folder } = data;
            return folder.listId ? (
              <Redirect to="design" />
            ) : (
              <SkipLogic {...props} />
            );
          }
        }}
      </Query>
    )}
  />,
];
