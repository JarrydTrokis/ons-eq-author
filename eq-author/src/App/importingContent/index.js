import React, { useState } from "react";
import { Query } from "react-apollo";
import { useParams, useLocation } from "react-router-dom";

import * as Headings from "constants/table-headings";

import GET_QUESTIONNAIRE_LIST from "graphql/getQuestionnaireList.graphql";
import GET_QUESTIONNAIRE from "graphql/getQuestionnaire.graphql";

import QuestionnairesView from "components/QuestionnairesView";
import QuestionnaireSelectModal from "components/modals/QuestionnaireSelectModal";
import ReviewQuestionsModal from "components/modals/ImportQuestionReviewModal";
import QuestionPicker from "components/QuestionPicker";

const SelectQuestionnaire = ({
  isOpen,
  questionnaires = [],
  onSelect,
  onCancel,
}) => {
  const [selectedQuestionnaire, setSelectedQuestionnaire] = useState(null);

  const enabledHeadings = [
    Headings.TITLE,
    Headings.OWNER,
    Headings.CREATED,
    Headings.MODIFIED,
  ];

  return (
    <QuestionnaireSelectModal
      isOpen={isOpen}
      onSelect={() => onSelect(selectedQuestionnaire)}
      onClose={onCancel}
      onCancel={onCancel}
      disableSelect={!selectedQuestionnaire}
    >
      <QuestionnairesView
        questionnaires={questionnaires}
        selectedQuestionnaire={selectedQuestionnaire}
        enabledHeadings={enabledHeadings}
        canCreateQuestionnaire={false}
        padding="small"
        onQuestionnaireClick={(questionnaireId) =>
          setSelectedQuestionnaire(
            questionnaires.find(({ id }) => id === questionnaireId)
          )
        }
      />
    </QuestionnaireSelectModal>
  );
};

const ImportingContent = ({ stopImporting }) => {
  /*
   * Modal display states
   */

  //True as it's the first modal in the process
  const [selectingQuestionnaire, setSelectingQuestionnaire] = useState(true);
  const [reviewingQuestions, setReviewingQuestions] = useState(false);
  const [selectingQuestions, setSelectingQuestions] = useState(false);

  /*
   * Data
   */

  const [questionnaireImportingFrom, setQuestionnaireImportingFrom] =
    useState(null);

  const [questionsToImport, setQuestionsToImport] = useState([]);

  /*
   * Handlers
   */

  const onSelectQuestionnaire = (questionnaire) => {
    setQuestionnaireImportingFrom(questionnaire);
    setSelectingQuestionnaire(false);
    setReviewingQuestions(true);
  };

  const onSelectQuestions = () => {
    setReviewingQuestions(false);
    setSelectingQuestions(true);
  };

  const onBackFromReviewingQuestions = () => {
    setReviewingQuestions(false);
    setSelectingQuestionnaire(true);
  };

  const onQuestionPickerCancel = () => {
    setSelectingQuestions(false);
    setReviewingQuestions(true);
  };

  const onQuestionPickerSubmit = (selection) => {
    setQuestionsToImport(selection);
    setSelectingQuestions(false);
    setReviewingQuestions(true);
  };

  const onGlobalCancel = () => {
    setReviewingQuestions(false);
    setSelectingQuestionnaire(false);
    setQuestionnaireImportingFrom(null);
    stopImporting();
  };

  const { questionnaireId: currentQuestionnaireId } = useParams();

  return (
    <>
      {selectingQuestionnaire && (
        <Query
          query={GET_QUESTIONNAIRE_LIST}
          variables={{
            input: { filter: { ne: { ids: [currentQuestionnaireId] } } },
          }}
        >
          {({ loading, error, data }) => {
            if (loading) {
              return <p>Loading</p>;
            }

            if (error || !data) {
              return <p>Error</p>;
            }

            return (
              <SelectQuestionnaire
                isOpen={selectingQuestionnaire}
                onCancel={onGlobalCancel}
                questionnaires={data.questionnaires}
                onSelect={onSelectQuestionnaire}
              />
            );
          }}
        </Query>
      )}
      {reviewingQuestions && (
        <ReviewQuestionsModal
          isOpen={reviewingQuestions}
          questionnaire={questionnaireImportingFrom}
          onCancel={onGlobalCancel}
          onBack={onBackFromReviewingQuestions}
          onSelectQuestions={onSelectQuestions}
          startingSelectedQuestions={questionsToImport}
        />
      )}
      {selectingQuestions && (
        <Query
          query={GET_QUESTIONNAIRE}
          variables={{
            input: { questionnaireId: questionnaireImportingFrom.id },
          }}
        >
          {({ loading, error, data }) => {
            if (loading) {
              return <p>Loading</p>;
            }

            if (error || !data) {
              return <p>Error</p>;
            }

            const { sections } = data.questionnaire;

            return (
              <QuestionPicker
                isOpen={selectingQuestions}
                sections={sections}
                startingSelectedQuestions={questionsToImport}
                warningPanel="You cannot import folders but you can import any questions they contain."
                showSearch
                onClose={onGlobalCancel}
                onCancel={onQuestionPickerCancel}
                onSubmit={onQuestionPickerSubmit}
              />
            );
          }}
        </Query>
      )}
    </>
  );
};

export default ImportingContent;
