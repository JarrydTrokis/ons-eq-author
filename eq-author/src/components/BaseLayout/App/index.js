import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { createGlobalStyle } from "styled-components";
import Theme from "contexts/themeContext";

import { lostConnection, gainConnection } from "redux/saving/actions";

const GlobalStyle = createGlobalStyle`
  html {
    box-sizing: border-box;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }

  html,
  body {
    height: 100%;
    width: 100%;
    font-size: 1em;
    margin: 0;
    padding: 0;
    -webkit-font-smoothing: antialiased;
  }

  body {
    font-family: 'Lato', sans-serif;
    overflow: hidden;
    color: ${({ theme }) => theme.colors.text};
  }

  input,
  select,
  textarea,
  button {
    font-family: inherit;
    -webkit-font-smoothing: inherit;
  }

  a {
    color: ${({ theme }) => theme.colors.blue};

    &:hover {
      text-decoration: none;
    }
  }
`;

export class UnconnectedApp extends React.Component {
  static propTypes = {
    children: PropTypes.element.isRequired,
    gainConnection: PropTypes.func.isRequired,
    lostConnection: PropTypes.func.isRequired,
  };

  componentDidMount = () => {
    window.addEventListener("online", this.props.gainConnection);
    window.addEventListener("offline", this.props.lostConnection);
  };

  componentWillUnmount = () => {
    window.removeEventListener("online", this.props.gainConnection);
    window.removeEventListener("offline", this.props.lostConnection);
  };

  render() {
    return (
      <Theme>
        {this.props.children} <GlobalStyle />
      </Theme>
    );
  }
}

export default connect(null, {
  lostConnection,
  gainConnection,
})(UnconnectedApp);
