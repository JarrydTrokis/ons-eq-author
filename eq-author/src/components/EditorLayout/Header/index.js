import styled from "styled-components";
import PropTypes from "prop-types";
import { flowRight } from "lodash";
import { withRouter } from "react-router-dom";
import gql from "graphql-tag";
import { useSubscription } from "react-apollo";

import CustomPropTypes from "custom-prop-types";

import { withQuestionnaire } from "components/QuestionnaireContext";

export const UtilityBtns = styled.div`
  display: flex;
  flex: 1 0 25%;
  justify-content: flex-end;
  margin-right: -1.5em;
`;

export const UnconnectedHeader = props => {
  const { match } = props;

  useSubscription(publishStatusSubscription, {
    variables: { id: match.params.questionnaireId },
  });
};

export const publishStatusSubscription = gql`
  subscription PublishStatus($id: ID!) {
    publishStatusUpdated(id: $id) {
      id
      publishStatus
    }
  }
`;

UnconnectedHeader.propTypes = {
  questionnaire: CustomPropTypes.questionnaire,
  title: PropTypes.string,
  client: PropTypes.shape({
    resetStore: PropTypes.func.isRequired,
  }),
  children: PropTypes.node,
  match: PropTypes.shape({
    params: PropTypes.shape({
      modifier: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default flowRight(withQuestionnaire, withRouter)(UnconnectedHeader);
