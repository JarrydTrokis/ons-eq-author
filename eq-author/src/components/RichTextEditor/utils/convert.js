import { convertToHTML, convertFromHTML } from "draft-convert";
import { EditorState } from "draft-js";

export const toHTML = entityMap => {
  const entityToHTML = (entity, originalText) => {
    const mapper = entityMap[entity.type];
    return mapper ? mapper(entity) : originalText;
  };

  const convert = convertToHTML({ entityToHTML });

  return editorState => convert(editorState.getCurrentContent());
};

export const fromHTML = nodeToFn => {
  const htmlToEntity = (nodeName, ...otherArgs) => {
    const entity = Object.entries(nodeToFn)
      .filter(([name]) => name === nodeName)
      .map(([, fn]) => fn(nodeName, ...otherArgs))
      .find(result => result);

    return entity || null;
  };

  const convert = convertFromHTML({ htmlToEntity });

  return (html, decorator) =>
    EditorState.createWithContent(convert(html), decorator);
};
