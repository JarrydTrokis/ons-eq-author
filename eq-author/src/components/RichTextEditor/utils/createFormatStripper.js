import { mapKeys } from "lodash";
import { filterEditorState } from "draftjs-filters";
import { filterConfig as pipedFilterConfig } from "components/RichTextEditor/entities/PipedValue";
import { filterConfig as linkFilterConfig } from "components/RichTextEditor/LinkPlugin";

const mapper = {
  bold: { format: "BOLD", type: "styles" },
  emphasis: { format: "ITALIC", type: "styles" },
  list: { format: "unordered-list-item", type: "blocks" },
  heading: { format: "header-two", type: "blocks" },
};

// Custom filter function to replace multiple spaces with a single space
function replaceMultipleSpaces(contentState) {
  const text = contentState.getText();
  const transformedText = text.replace(/\s+/g, " "); // Replace multiple spaces with a single space
  return contentState.merge({
    text: transformedText,
  });
}

export default function createFormatStripper(controls) {
  const filterConfiguration = {
    blocks: [],
    styles: [],
    entities: [pipedFilterConfig, linkFilterConfig],
    maxNesting: 0,
    whitespacedCharacters: [replaceMultipleSpaces],
  };

  mapKeys(controls, (value, key) => {
    if (mapper[key] && controls[key]) {
      filterConfiguration[mapper[key].type].push(mapper[key].format);
    }
  });

  return (editorState) => filterEditorState(filterConfiguration, editorState);
}
