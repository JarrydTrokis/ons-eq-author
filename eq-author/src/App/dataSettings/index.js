import React from "react";
import { Route, Redirect } from "react-router-dom";
import SampleFileDataPage from "./SampleFileData";
import SupplementaryDataPage from "./SupplementaryDataPage";
import { buildDataPath } from "utils/UrlUtils";

export default [
  <Route
    key="sample"
    exact
    path="/q/:questionnaireId/data/sample-file-data"
    component={SampleFileDataPage}
  />,
  <Route
    key="dataset"
    exact
    path="/q/:questionnaireId/data/supplementary-data"
    component={SupplementaryDataPage}
  />,
  <Route
    key="data"
    exact
    path="/q/:questionnaireId/data"
    render={(props) => (
      <Redirect to={`${buildDataPath(props.match.params)}/sample-file-data`} />
    )}
  />,
];
