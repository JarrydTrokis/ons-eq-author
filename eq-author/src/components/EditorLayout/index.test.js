import React from "react";
import { shallow } from "enzyme";
import { byTestAttr } from "tests/utils/selectors";
import { MeContext } from "App/MeContext";

import EditorLayout from "./";

describe("Editor Layout", () => {
  let props;
  beforeEach(() => {
    props = {
      page: {
        id: "1",
        displayName: "Foo",
        answers: [],
      },
      design: true,
      preview: false,
      routing: true,
    };
  });

  const user = {
    id: "1",
    name: "Name",
  };

  it("should render", () => {
    const wrapper = shallow(
      <MeContext.Provider value={{ me: user }}>
        <EditorLayout {...props}>Content</EditorLayout>
      </MeContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should add new question page to correct section", () => {
    const onAddQuestionPage = jest.fn();

    const wrapper = shallow(
      <MeContext.Provider value={{ me: user }}>
        <EditorLayout {...props} onAddQuestionPage={onAddQuestionPage}>
          Content
        </EditorLayout>
      </MeContext.Provider>
    );

    wrapper.find(byTestAttr("btn-add-page")).first().simulate("click");

    expect(onAddQuestionPage).toHaveBeenCalled();
  });

  it("should render the panel when a render function is provided", () => {
    const renderPanel = jest
      .fn()
      .mockReturnValue(<div data-test="test-panel">Panel</div>);

    const wrapper = shallow(
      <MeContext.Provider value={{ me: user }}>
        <EditorLayout
          {...props}
          singleColumnLayout={null}
          renderPanel={renderPanel}
        >
          Content
        </EditorLayout>
      </MeContext.Provider>
    );

    expect(renderPanel).toHaveBeenCalled();
    expect(wrapper.find("[data-test='test-panel']")).toHaveLength(1);
  });

  it("should render right hand panel if singleColumnLayout is null", () => {
    const wrapper = shallow(
      <MeContext.Provider value={{ me: user }}>
        <EditorLayout {...props}>Content</EditorLayout>
      </MeContext.Provider>
    );
    expect(wrapper.find("[data-test='right-hand-panel']")).toHaveLength(1);
  });

  it("should NOT render right hand panel if singleColumnLayout is true", () => {
    const wrapper = shallow(
      <MeContext.Provider value={{ me: user }}>
        <EditorLayout {...props} singleColumnLayout>
          Content
        </EditorLayout>
      </MeContext.Provider>
    );

    expect(wrapper.find("[data-test='right-hand-panel']")).toHaveLength(0);
  });
});
