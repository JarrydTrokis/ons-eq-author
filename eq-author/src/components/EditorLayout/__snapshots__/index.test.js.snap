// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Layout should render 1`] = `
<GetContext
  title={[Function]}
>
  <EditorLayout__Container>
    <Component>
      <withRouter(UnwrappedTabs)
        design={true}
        preview={false}
      />
    </Component>
    <ScrollPane
      scrollToTop={true}
    >
      <EditorLayout__StyledGrid
        align="top"
        fillHeight={true}
        page={
          Object {
            "answers": Array [],
            "displayName": "Foo",
            "id": "1",
          }
        }
        routing={true}
      >
        <Column
          cols={12}
          gutters={false}
        >
          <EditorLayout__Margin>
            <MainCanvas
              maxWidth="70em"
            >
              Content
              <EditorLayout__PanelWrapper
                data-test="property-panel"
              />
            </MainCanvas>
          </EditorLayout__Margin>
        </Column>
      </EditorLayout__StyledGrid>
    </ScrollPane>
  </EditorLayout__Container>
</GetContext>
`;
